File: js/src/sources.js
// ╔═══════════════════════════════════════════════════════╗
// ║                                                       ║
// ║                 Hydra de Lerne                        ║
// ║               All rights reserved                     ║
// ║                  onvo.me/hydra                        ║
// ║                                                       ║
// ╚═══════════════════════════════════════════════════════╝


// for nodejs <= 18 un comment this

// let fetch;

// (async () => {
//     fetch = (await import('node-fetch')).default;
// })();

const generateClientPlaybackNonce = (length) => {
    const CPN_CHARS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_';
    return Array.from({ length }, () => CPN_CHARS[Math.floor(Math.random() * CPN_CHARS.length)]).join('');
};

const currentVersion = 1.0

const iosPayload = {
    data: () => {
        return {
            cpn: generateClientPlaybackNonce(16),
            contentCheckOk: true,
            racyCheckOk: true,
            context: {
                client: {
                    clientName: 'IOS',
                    clientVersion: '19.228.1',
                    deviceMake: 'Apple',
                    deviceModel: 'iPhone16,2',
                    platform: 'MOBILE',
                    osName: 'iOS',
                    osVersion: '17.5.1.21F90',
                    hl: 'en',
                    gl: 'US',
                    utcOffsetMinutes: -240,
                },
                request: {
                    internalExperimentFlags: [],
                    useSsl: true,
                },
                user: {
                    lockedSafetyMode: false,
                },
            }
        }
    },
    agent: `com.google.ios.youtube/19.228.1(iPhone16,2; U; CPU iOS 17_5_1 like Mac OS X; en_US)`
}

const remotePaylod = {
    data: () => {
        return {
            cpn: generateClientPlaybackNonce(16),
            contentCheckOk: true,
            racyCheckOk: true,
            context: {
                client: {
                    clientName: 'ANDROID',
                    clientVersion: '19.30.36',
                    platform: 'MOBILE',
                    osName: 'Android',
                    osVersion: '14',
                    androidSdkVersion: '34',
                    hl: 'en',
                    gl: 'US',
                    utcOffsetMinutes: -240,
                },
                request: {
                    internalExperimentFlags: [],
                    useSsl: true,
                },
                user: {
                    lockedSafetyMode: false,
                },
            }
        }
    },
    agent: `com.google.android.youtube/19.30.36 (Linux; U; Android 14; en_US) gzip`
}

export const getData = async (videoId, isRaw, payloadData) => {
    const payload = {
        videoId,
        ...payloadData || remotePaylod.data()
    };

    const query = {
        prettyPrint: false,
        t: generateClientPlaybackNonce(12),
        id: videoId,
    };

    const queryString = new URLSearchParams(query).toString();
    const opts = {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            'User-Agent': remotePaylod.agent,
            'X-Goog-Api-Format-Version': '2',
        },
        body: JSON.stringify(payload),
    };
    let data = {}
    try {
        const response = await fetch(`https://youtubei.googleapis.com/youtubei/v1/player?${queryString}`, opts);
        if (!response.ok) {
            hasError = true
            checkUpdate()
        }
        data = await response.json()
    } catch (e) {

    }
    return !isRaw ? parseFormats(data) : data;
};

let hasError = false

export const checkUpdate = async () => {
    try {
        const response = await fetch('https://st.onvo.me/config.json')
        const json = await response.json()
        if (json.version !== currentVersion) {
            if (json.data && hasError || json.forceUpdate) {
                remotePaylod.data = () => {
                    return {
                        cpn: generateClientPlaybackNonce(16),
                        ...json.data
                    }
                }
            }
            if(json.agent && hasError || json.forceUpdate){
                remotePaylod.agent = json.agent
            }
        }
    } catch (e) {

    }
}

const parseFormats = (response) => {
    let formats = [];
    if (response && response.streamingData) {
        formats = formats
            .concat(response.streamingData.formats || [])
            .concat(response.streamingData.adaptiveFormats || []);
    }
    return formats;
};

export const filter = (formats, filter, options = {}) => {
    if (!Array.isArray(formats)) {
        throw new Error('Formats must be an array');
    }

    if (!filter) {
        throw new Error('Filter must be provided');
    }

    const {
        fallback = false,
        customSort = null,
        minBitrate = 0,
        minResolution = 0,
        codec = null,
    } = options;

    let fn;

    const filterByCodec = (format) => {
        if (!codec) return true;
        return format.mimeType.includes(codec);
    };

    const filterByBitrate = (format) => {
        return (format.bitrate || 0) >= minBitrate;
    };

    const filterByResolution = (format) => {
        return (format.width || 0) >= minResolution || (format.height || 0) >= minResolution;
    };

    const filterByUrl = (format) => !!format.url;

    const applyFilters = (format) => {
        return filterByUrl(format) && filterByCodec(format) && filterByBitrate(format) && filterByResolution(format);
    };

    switch (filter) {
        case 'bestvideo':
            fn = (format) => format.mimeType.includes('video');
            return formats
                .filter((format) => applyFilters(format) && fn(format))
                .sort(customSort || ((a, b) => (b.width || 0) - (a.width || 0) || (b.bitrate || 0) - (a.bitrate || 0)))[0];

        case 'bestaudio':
            fn = (format) => format.mimeType.includes('audio');
            return formats
                .filter((format) => applyFilters(format) && fn(format))
                .sort(customSort || ((a, b) => (b.bitrate || 0) - (a.bitrate || 0)))[0];

        case 'lowestvideo':
            fn = (format) => format.mimeType.includes('video');
            return formats
                .filter((format) => applyFilters(format) && fn(format))
                .sort(customSort || ((a, b) => (a.width || 0) - (b.width || 0) || (a.bitrate || 0) - (b.bitrate || 0)))[0];

        case 'lowestaudio':
            fn = (format) => format.mimeType.includes('audio');
            return formats
                .filter((format) => applyFilters(format) && fn(format))
                .sort(customSort || ((a, b) => (a.bitrate || 0) - (b.bitrate || 0)))[0];

        case 'videoandaudio':
        case 'audioandvideo':
            fn = (format) => format.mimeType.includes('video') && format.mimeType.includes('audio');
            break;

        case 'video':
            fn = (format) => format.mimeType.includes('video');
            break;

        case 'videoonly':
            fn = (format) => format.mimeType.includes('video') && !format.mimeType.includes('audio');
            break;

        case 'audio':
            fn = (format) => format.mimeType.includes('audio');
            break;

        case 'audioonly':
            fn = (format) => !format.mimeType.includes('video') && format.mimeType.includes('audio');
            break;

        default:
            if (typeof filter === 'function') {
                fn = filter;
            } else if (typeof filter === 'string' && filter.startsWith('extension:')) {
                const extension = filter.split(':')[1];
                fn = (format) => format.url.includes(`.${extension}`);
            } else {
                throw new Error(`Given filter (${filter}) is not supported`);
            }
    }

    const filteredFormats = formats.filter((format) => applyFilters(format) && fn(format));

    if (fallback && filteredFormats.length === 0) {
        return formats.filter(filterByUrl)[0];
    }

    return filteredFormats;
};

File: python/hydra_youtube_api/sources.py
import json
import random
import string
from urllib.parse import urlencode
import requests

IOS_CLIENT_VERSION = '19.28.1'
IOS_DEVICE_MODEL = 'iPhone16,2'
IOS_USER_AGENT_VERSION = '17_5_1'
IOS_OS_VERSION = '17.5.1.21F90'

ANDROID_CLIENT_VERSION = '19.30.36'
ANDROID_OS_VERSION = '14'
ANDROID_SDK_VERSION = '34'

def generate_client_playback_nonce(length):
    CPN_CHARS = string.ascii_letters + string.digits + '-_'
    return ''.join(random.choice(CPN_CHARS) for _ in range(length))

async def get_data(video_id, is_raw=False, client_name='ios'):
    client = {}
    agent = f'com.google.ios.youtube/{IOS_CLIENT_VERSION}({IOS_DEVICE_MODEL}; U; CPU iOS {IOS_USER_AGENT_VERSION} like Mac OS X; en_US)'

    if client_name == 'ios':
        client = {
            'clientName': 'IOS',
            'clientVersion': IOS_CLIENT_VERSION,
            'deviceMake': 'Apple',
            'deviceModel': IOS_DEVICE_MODEL,
            'platform': 'MOBILE',
            'osName': 'iOS',
            'osVersion': IOS_OS_VERSION,
            'hl': 'en',
            'gl': 'US',
            'utcOffsetMinutes': -240,
        }
    elif client_name == 'android':
        client = {
            'clientName': 'ANDROID',
            'clientVersion': ANDROID_CLIENT_VERSION,
            'platform': 'MOBILE',
            'osName': 'Android',
            'osVersion': ANDROID_OS_VERSION,
            'androidSdkVersion': ANDROID_SDK_VERSION,
            'hl': 'en',
            'gl': 'US',
            'utcOffsetMinutes': -240,
        }
        agent = f'com.google.android.youtube/{ANDROID_CLIENT_VERSION} (Linux; U; Android {ANDROID_OS_VERSION}; en_US) gzip'

    payload = {
        'videoId': video_id,
        'cpn': generate_client_playback_nonce(16),
        'contentCheckOk': True,
        'racyCheckOk': True,
        'context': {
            'client': client,
            'request': {
                'internalExperimentFlags': [],
                'useSsl': True,
            },
            'user': {
                'lockedSafetyMode': False,
            },
        },
    }

    query = {
        'prettyPrint': False,
        't': generate_client_playback_nonce(12),
        'id': video_id,
    }

    query_string = urlencode(query)
    headers = {
        'Content-Type': 'application/json',
        'User-Agent': agent,
        'X-Goog-Api-Format-Version': '2',
    }

    response = requests.post(
        f'https://youtubei.googleapis.com/youtubei/v1/player?{query_string}',
        headers=headers,
        data=json.dumps(payload)
    )

    data = response.json()
    return data if is_raw else parse_formats(data)

def parse_formats(response):
    formats = []
    if response and 'streamingData' in response:
        formats.extend(response['streamingData'].get('formats', []))
        formats.extend(response['streamingData'].get('adaptiveFormats', []))
    return formats
# hydra_youtube_api/sources.py
def filter_formats(formats, filter_type = 'all', options=None):
    if options is None:
        options = {}

    if not isinstance(formats, list):
        raise ValueError('Formats must be a list')

    if not filter_type:
        raise ValueError('Filter must be provided')

    fallback = options.get('fallback', False)
    custom_sort = options.get('customSort', None)
    min_bitrate = options.get('minBitrate', 0)
    min_resolution = options.get('minResolution', 0)
    codec = options.get('codec', None)

    def filter_by_codec(format):
        if not codec:
            return True
        return codec in format.get('mimeType', '')

    def filter_by_bitrate(format):
        return format.get('bitrate', 0) >= min_bitrate

    def filter_by_resolution(format):
        return format.get('width', 0) >= min_resolution or format.get('height', 0) >= min_resolution

    def filter_by_url(format):
        return 'url' in format

    def apply_filters(format):
        return filter_by_url(format) and filter_by_codec(format) and filter_by_bitrate(format) and filter_by_resolution(format)

    if filter_type == 'all':
        # Return all formats that pass the filters
        return [format for format in formats if apply_filters(format)]

    elif filter_type == 'bestvideo':
        fn = lambda format: 'video' in format.get('mimeType', '')
        filtered = [format for format in formats if apply_filters(format) and fn(format)]
        if custom_sort:
            filtered.sort(key=custom_sort)
        else:
            filtered.sort(key=lambda x: (-x.get('width', 0), -x.get('bitrate', 0)))
        return filtered[0] if filtered else None

    elif filter_type == 'bestaudio':
        fn = lambda format: 'audio' in format.get('mimeType', '')
        filtered = [format for format in formats if apply_filters(format) and fn(format)]
        if custom_sort:
            filtered.sort(key=custom_sort)
        else:
            filtered.sort(key=lambda x: -x.get('bitrate', 0))
        return filtered[0] if filtered else None

    elif filter_type == 'lowestvideo':
        fn = lambda format: 'video' in format.get('mimeType', '')
        filtered = [format for format in formats if apply_filters(format) and fn(format)]
        if custom_sort:
            filtered.sort(key=custom_sort)
        else:
            filtered.sort(key=lambda x: (x.get('width', 0), x.get('bitrate', 0)))
        return filtered[0] if filtered else None

    elif filter_type == 'lowestaudio':
        fn = lambda format: 'audio' in format.get('mimeType', '')
        filtered = [format for format in formats if apply_filters(format) and fn(format)]
        if custom_sort:
            filtered.sort(key=custom_sort)
        else:
            filtered.sort(key=lambda x: x.get('bitrate', 0))
        return filtered[0] if filtered else None

    elif filter_type in ['videoandaudio', 'audioandvideo']:
        fn = lambda format: 'video' in format.get('mimeType', '') and 'audio' in format.get('mimeType', '')
    elif filter_type == 'video':
        fn = lambda format: 'video' in format.get('mimeType', '')
    elif filter_type == 'videoonly':
        fn = lambda format: 'video' in format.get('mimeType', '') and 'audio' not in format.get('mimeType', '')
    elif filter_type == 'audio':
        fn = lambda format: 'audio' in format.get('mimeType', '')
    elif filter_type == 'audioonly':
        fn = lambda format: 'audio' in format.get('mimeType', '') and 'video' not in format.get('mimeType', '')
    else:
        if callable(filter_type):
            fn = filter_type
        elif isinstance(filter_type, str) and filter_type.startswith('extension:'):
            extension = filter_type.split(':')[1]
            fn = lambda format: f'.{extension}' in format.get('url', '')
        else:
            raise ValueError(f'Given filter ({filter_type}) is not supported')

    filtered_formats = [format for format in formats if apply_filters(format) and fn(format)]

    if fallback and not filtered_formats:
        return next((format for format in formats if filter_by_url(format)), None)

    return filtered_formats